### PR TITLE
Enable px4io on pi 2

### PR DESCRIPTION
**Pull request by jlaitine**
Enable px4io

I will still optimize the IRQ load by using rx buffer threshold interrupt, when the threshold is freely settable in the fpga uart block

### CODE REVIEW

File: platforms/nuttx/src/px4/microchip/microchip_common/include/px4_arch/px4io_serial.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	perf_counter_t		_pc_txns;
+	perf_counter_t		_pc_retries;
+	perf_counter_t		_pc_timeouts;
+	perf_counter_t		_pc_crcerrs;
+	perf_counter_t		_pc_protoerrs;
+	perf_counter_t		_pc_uerrs;
+	perf_counter_t		_pc_idle;
+	perf_counter_t		_pc_badidle;
+private:
+	/*
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/microchip/microchip_common/include/px4_arch/px4io_serial.h:86]: this comment is copy-paste from pixhawk code, and nonsense...

File: platforms/nuttx/src/px4/microchip/microchip_common/include/px4_arch/px4io_serial.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	 * Packet overhead is 26µs for the four-byte header.
+	 *
+	 * 32 registers = 451µs
+	 *
+	 * Maybe we can just send smaller packets (e.g. 8 regs) and loop for larger (less common)
+	 * transfers? Could cause issues with any regs expecting to be written atomically...
+	 */
+	IOPacket		*_io_buffer_ptr;
+
+	/** bus-ownership lock */
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/microchip/microchip_common/include/px4_arch/px4io_serial.h:99]: fix comment

**Comment by jlaitine** [comment]
Removed extra mutual exclusion, it was already managed in src/drivers/px4io common code, no need to do it in _bus_exchange.

**Comment by jlaitine** [comment]
Still fixed the uart baud rate calculation. It was a copy paste from mpfs_serial.c, and looked very strange to me...

**Comment by jlaitine** [comment]
Added critical section around uart tx fifo fill; the px4io expects to get the bytes in one burst. A break in there will timeout the px4io dma and cause error

File: platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+	uint32_t lcr = UART_LCR_DLS_8BITS;
+
+	/* Set bits, enter dlab */
+
+	putreg32(lcr | UART_LCR_DLAB, PX4IO_SERIAL_BASE + MPFS_UART_LCR_OFFSET);
+
+	/*  Baud_value = PCLK_Frequency / (baud_rate * 16)
+	 *
+	 *  Calculate in fixed point 28.6 format (fraction is presented by 6 bits in PF)
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp:112]: this comment is wrong. it is 26.6 :)

File: platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+	/* Get the current status */
+
+	uint32_t status = getreg32(PX4IO_SERIAL_BASE + MPFS_UART_IIR_OFFSET);
+	uint8_t *packet = (uint8_t *)_current_packet;
+	bool first_rcv = _packet_cnt == 0 ? true : false;
+	int ret = OK;
+
+	switch (status & UART_IIR_IID_MASK) {
+
+	case UART_IIR_IID_RECV:
```
Comment (pussuw) [inline platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp:332]: Where is the interrupt cleared or is this level active, i.e. when FIFO is below threshold IRQ line is deasserted ?

File: platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+		instance->_do_interrupt();
+	}
+
+	return 0;
+}
+
+void
+ArchPX4IOSerial::_do_interrupt()
+{
+
```
Comment (pussuw) [inline platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp:322]: If you have interest you could remove these extra line feeds. Makes no difference to me. ODC might trigger for some.

File: platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+		break;
+	}
+
+	return ret;
+}
+
+int
+ArchPX4IOSerial::_bus_exchange(IOPacket *_packet)
+{
+
```
Comment (pussuw) [inline platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp:235]: If you have interest you could remove these extra line feeds. Makes no difference to me. ODC might trigger for some.

File: platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+ArchPX4IOSerial::ArchPX4IOSerial() :
+	_current_packet(nullptr),
+	_completion_semaphore(SEM_INITIALIZER(0))
+{
+}
+
+ArchPX4IOSerial::~ArchPX4IOSerial()
+{
+
```
Comment (pussuw) [inline platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp:62]: If you have interest you could remove these extra line feeds. Makes no difference to me. OCD might trigger for some.

**Review by pussuw — APPROVED** [review]
LGTM, the IRQ concern is most likely not an issue

**Pull request merged**